ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,240 @@
+    y_coordinates: GenericArray<F, N>,
+}
+
+/// The Canonical Lagrange denominator is defined as the denominator of the Lagrange base polynomials
+/// `https://en.wikipedia.org/wiki/Lagrange_polynomial`
+/// where the "x coordinates" of the input points are `x_0` to `x_N` are `F::ZERO` to `(N-1)*F::ONE`
+/// the degree of the polynomials is `N-1`
+pub struct CanonicalLagrangeDenominator<F: Field, N: ArrayLength> {
+    denominator: GenericArray<F, N>,
+}
+
+impl<F, N> CanonicalLagrangeDenominator<F, N>
+where
+    F: PrimeField,
+    N: ArrayLength,
+{
+    /// generates canonical Lagrange denominators
+    ///
+    /// ## Panics
+    /// When the field size is too small for `N` evaluation points
+    pub fn new() -> Self {
+        // assertion that field is large enough
+        // also checks that `try_from` for conversions from sufficiently small `u128` to `F` do not panic
+        debug_assert!(F::BITS > usize::BITS - N::USIZE.leading_zeros());
+
+        let mut denominator = iter::repeat(F::ONE)
+            .take(N::USIZE)
+            .collect::<GenericArray<_, N>>();
+        for (i, d) in denominator.iter_mut().enumerate() {
+            for j in (0usize..N::USIZE).filter(|&j| i != j) {
+                *d *= F::try_from(i as u128).unwrap() - F::try_from(j as u128).unwrap();
+            }
+            *d = d.invert();
+        }
+        CanonicalLagrangeDenominator { denominator }
+    }
+}
+
+/// `LagrangeTable` is a precomputation table for the Lagrange evaluation.
+/// The "x coordinates" of the input points are `x_0` to `x_(N-1)` are `F::ZERO` to `(N-1)*F::ONE`.
+/// The `LagrangeTable` also specifies `M` "x coordinates" for the output points
+/// The "x coordinates" of the output points are `x_N` to `x_(N+M-1)` are `N*F::ONE` to `(N+M-1)*F::ONE`.
+pub struct LagrangeTable<F: Field, N: ArrayLength, M: ArrayLength> {
+    table: GenericArray<GenericArray<F, N>, M>,
+}
+
+impl<F, N> LagrangeTable<F, N, U1>
+where
+    F: Field,
+    N: ArrayLength,
+{
+    /// generates a `CanonicalLagrangeTable` from `CanoncialLagrangeDenominators` for a single output point
+    /// The "x coordinate" of the output point is `x_output`.
+    pub fn new(denominator: &CanonicalLagrangeDenominator<F, N>, x_output: &F) -> Self {
+        let mut table = denominator.denominator.clone();
+        compute_table_row(x_output, &mut table);
+        LagrangeTable::<F, N, U1> {
+            table: GenericArray::from_array([table; 1]),
+        }
+    }
+}
+
+impl<F, N, M> LagrangeTable<F, N, M>
+where
+    F: Field,
+    N: ArrayLength,
+    M: ArrayLength,
+{
+    /// This function uses the `LagrangeTable` to evaluate `polynomial` on the specified output "x coordinates"
+    /// outputs the "y coordinates" such that `(x,y)` lies on `polynomial`
+    pub fn eval(&self, polynomial: &Polynomial<F, N>) -> GenericArray<F, M> {
+        let mut result = iter::repeat(F::ONE)
+            .take(M::USIZE)
+            .collect::<GenericArray<_, _>>();
```

Comment:
Can't this use:

```suggestion
        let mut result = arr![F::ONE; M::USIZE];
```

or maybe:


```suggestion
        let mut result = arr![F::ONE; M];
```